tdms/include/arrays.h

```diff
@@ -0,0 +1,203 @@
+    switch (c) {
+      case 'x': return x;
+      case 'y': return y;
+      case 'z': return z;
+      default: throw std::runtime_error("Have no element " + to_string(c));
+    }
+  }
+};
+
+class XYZVectors {
+public:
+  double* x = nullptr;
+  double* y = nullptr;
+  double* z = nullptr;
+
+  /**
+   * Default constructor
+   */
+  XYZVectors() = default;
+
+  /**
+   * Set the pointer for one of the vectors in this collection with a name of c
+   * @param c Character labeling the vector
+   * @param ptr Pointer to assign
+   */
+  void set_ptr(char c, double* ptr);
+};
+
+// TODO: docstring
+class MaterialCollection{
+protected:
+  static void init_xyz_vectors(const mxArray *ptr, XYZVectors &arrays, const std::string &prefix);
+};
+
+class CCollectionBase {
+public:
+  XYZVectors a;
+  XYZVectors b;
+  XYZVectors c;
+};
+
+// TODO: docstring
+class CCollection : public CCollectionBase {
+private:
+  void init_xyz_vectors(const mxArray *ptr, XYZVectors &arrays, const std::string &prefix);
+
+public:
+  bool is_multilayer = false;
+  bool is_disp_ml = false;
+
+  explicit CCollection(const mxArray *ptr);
+};
+
+// TODO: docstring
+class CMaterial : public CCollectionBase, MaterialCollection {
+public:
+  explicit CMaterial(const mxArray *ptr);
+};
+
+class DCollectionBase {
+public:
+  XYZVectors a;
+  XYZVectors b;
+};
+
+// TODO: docstring
+class DCollection: public DCollectionBase{
+private:
+  static void init_xyz_vectors(const mxArray *ptr, XYZVectors &arrays, const std::string &prefix);
+
+public:
+  explicit DCollection(const mxArray *ptr);
+};
+
+// TODO: docstring
+class DMaterial : public DCollectionBase, MaterialCollection {
+public:
+  explicit DMaterial(const mxArray *ptr);
+};
+
+class DispersiveMultiLayer{
+public:
+  double* alpha = nullptr;
+  double* beta = nullptr;
+  double* gamma = nullptr;
+  XYZVectors kappa;
+  XYZVectors sigma;
+
+public:
+  explicit DispersiveMultiLayer(const mxArray *ptr);
+};
+
+template<typename T>
+class Matrix{
+protected:
+  int n_rows = 0;
+  int n_cols = 0;
+  T** matrix = nullptr;
+
+public:
+  inline T* operator[] (int value) const { return matrix[value]; }
+
+  bool has_elements(){ return matrix != nullptr; };
+};
+
+class GratingStructure: public Matrix<int>{
+
+public:
+  GratingStructure(const mxArray *ptr, int I_tot);
+
+  ~GratingStructure();
+};
+
+template<typename T>
+class Vector{
+protected:
+  int n = 0;        // Number of elements
+  T* vector = nullptr; // Internal array
+
+public:
+  Vector() = default;
+
+  explicit Vector(const mxArray *ptr);
+
+  inline T operator[] (int value) const { return vector[value]; };
+
+  inline int size() const{ return n; };
+};
+
+class FrequencyExtractVector: public Vector<double>{
+public:
+  FrequencyExtractVector(const mxArray *ptr, double omega_an);
+};
+
+class FrequencyVectors{
+public:
+  Vector<double> x;
+  Vector<double> y;
+
+  void initialise(const mxArray *ptr);
+};
+
+class Pupil: public Matrix<double>{
```

Comment:
I'm a bit lost as to what this actually is.... really a →👁️  ?

Reply:
no idea 😞  – hence the lack of docstring – another one for Peter